airflow/contrib/kubernetes/pod_launcher.py

```diff
@@ -0,0 +1,121 @@
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+import json
+import time
+from airflow.utils.log.logging_mixin import LoggingMixin
+from airflow.utils.state import State
+from datetime import datetime as dt
+from airflow.contrib.kubernetes.kubernetes_request_factory import \
+    pod_request_factory as pod_fac
+from kubernetes import watch
+from kubernetes.client.rest import ApiException
+from airflow import AirflowException
+
+from .kube_client import get_kube_client
+
+
+class PodStatus(object):
+    PENDING = 'pending'
+    RUNNING = 'running'
+    FAILED = 'failed'
+    SUCCEEDED = 'succeeded'
+
+
+class PodLauncher(LoggingMixin):
+    def __init__(self, kube_client=None):
+        super(PodLauncher, self).__init__()
+        self._client = kube_client or get_kube_client()
+        self._watch = watch.Watch()
+        self.kube_req_factory = pod_fac.SimplePodRequestFactory()
+
+    def run_pod_async(self, pod):
+        req = self.kube_req_factory.create(pod)
+        self.log.debug('Pod Creation Request: \n{}'.format(json.dumps(req, indent=2)))
+        try:
+            resp = self._client.create_namespaced_pod(body=req, namespace=pod.namespace)
+            self.log.debug('Pod Creation Response: {}'.format(resp))
+        except ApiException:
+            self.log.exception('Exception when attempting to create Namespaced Pod.')
+            raise
+        return resp
+
+    def run_pod(self, pod, startup_timeout=120):
+        # type: (Pod) -> State
+        """
+        Launches the pod synchronously and waits for completion.
+
+        Args:
+            pod (Pod):
+            startup_timeout (int): Timeout for startup of the pod (if pod is pending for
+             too long, considers task a failure
+        """
+        resp = self.run_pod_async(pod)
+        curr_time = dt.now()
+        if resp.status.start_time is None:
+            while self.pod_not_started(pod):
+                delta = dt.now() - curr_time
+                if delta.seconds >= startup_timeout:
+                    raise AirflowException("Pod took too long to start")
+                time.sleep(1)
+            self.log.debug('Pod not yet started')
+
+        final_status = self._monitor_pod(pod)
+        return final_status
+
+    def _monitor_pod(self, pod):
+        # type: (Pod) -> State
+
+        while self.pod_is_running(pod):
+            self.log.info("Pod {} has state {}".format(pod.name, State.RUNNING))
```

Comment:
Would it make sense to pass logs from the pod along to airflow, so it appears in the task logs? Eg. so you can track the progress of the worker pod in the airflow web UI.

Reply:
@kevincvlam good point. Should be pretty easy to use the kubernetes client to live-log the results of the pod. This would also handle the logging issues for cases where people don't have ELK (though we probably want this to be optional in case people have their own logging solutions)

Reply:
@kevincvlam That's a good question and there are a bunch of potential solutions. (You can use a read/write many persistent volume and mount the logs folder into each worker, you can intercept the logs from the kubernetes api itself, you can use a custom airflow logging solution). 

All of them have tradeoffs, I think we chose not to do anything because there were so many options and not a clear winner. I'm open to exploring this more as I think it's especially important for the executor.